### PR TITLE
Fix access-list address adds and deletes

### LINKED ISSUE #53: TERRA-45 ⁃ astra_access_list only creates the first address
### Terraform Version
Terraform core: `1.0.11`
Astra provider: `1.0.12`

### Affected Resource(s)
Please list the resources as a list, for example:

* astra_access_list

### Terraform Configuration Files
```hcl
module "astra" {
  ...
  private_link_access_cidr = [
    "10.0.0.1/32",
    "10.0.0.2/32",
    "10.0.0.3/32",
  ]
}

# actual module
variable "private_link_access_cidr" {
  type        = list(string)
  description = "List of CIDR's with access to the private link"
  default     = []
}

resource "astra_database" "database" {
  ...
}

resource "astra_access_list" "access_list" {
  database_id = astra_database.database.id

  addresses {
    dynamic "request" {
      for_each = var.private_link_access_cidr
      content {
        address = request.value
        enabled = true
      }
    }
  }
}

```

### Expected Behavior
The database should have an access list with 3 records.

### Actual Behavior
The terraform plan shows the resource as it is intended:

```
# module.astra.astra_access_list.access_list will be created
  + resource "astra_access_list" "access_list" {
      + database_id = (known after apply)
      + id          = (known after apply)

      + addresses {
          + request {
              + address = "10.0.0.1/32"
              + enabled = true
            }
          + request {
              + address = "10.0.0.2/32"
              + enabled = true
            }
          + request {
              + address = "10.0.0.3/32"
              + enabled = true
            }
        }
    }
```

Only the first address is created. See screenshot below.

image

The tfstate contains the 3 records:

```
    {
      "module": "module.astra",
      "mode": "managed",
      "type": "astra_access_list",
      "name": "access_list",
      "provider": "provider[\"registry.terraform.io/datastax/astra\"]",
      "instances": [
        {
          "schema_version": 0,
          "attributes": {
            "addresses": [
              {
                "request": [
                  {
                    "address": "10.0.0.1/32",
                    "description": "",
                    "enabled": true
                  },
                  {
                    "address": "10.0.0.2/32",
                    "description": "",
                    "enabled": true
                  },
                  {
                    "address": "10.0.0.3/32",
                    "description": "",
                    "enabled": true
                  }
                ]
              }
            ],
            "database_id": "*****************",
            "id": "*******************"
          },
          "sensitive_attributes": [],
          "private": "bnVsbA==",
          "dependencies": [
            "module.astra.astra_database.database"
          ]
        }
      ]
    },
```

### Steps to Reproduce
Just apply the code on a new database.

┆Issue is synchronized with this [Jira Task](https://datastax-oss.atlassian.net/browse/TERRA-45) by [Unito](https://www.unito.io)
┆friendlyId: TERRA-45
┆priority: Major

### PR DESCRIPTION
**Pull request by emerkle826**
Fixes #53

### CODE REVIEW

File: internal/provider/resource_access_list_test.go [hunk 1]
```diff
@@ -27,7 +26,15 @@ resource "astra_access_list" "example" {
   database_id = "aba3cf20-d579-4091-a36d-9c9f75096031"
   addresses {
     request {
-      address= "0.0.0.0/0"
+      address= "0.0.0.1/0"
```
Comment (jeffbanks) [inline internal/provider/resource_access_list_test.go:29]: Is it worth some additional coverage to add a `not enabled` to the test mix?

File: internal/provider/resource_access_list.go [hunk 2]
```diff
@@ -124,17 +123,31 @@ func resourceAccessListDelete(ctx context.Context, d *schema.ResourceData, meta
...
+	// 	&addressesQP,
+	// }
+	// client.DeleteAddressesOrAccessListForDatabase(ctx, astra.DatabaseIdParam(databaseID), params)
 
-	params :=  &astra.DeleteAddressesOrAccessListForDatabaseParams{
-		&addressesQP,
+	// The above code should work, but Astra seems to only delete the first address passed as a query param
+	// Until it's fixed in Astra, call DELETE for each address
+	aResp := *accessList.Addresses
+	if len(aResp) > 0 {
```
Comment (jeffbanks) [inline internal/provider/resource_access_list.go:141]: Formatting looks to be off a bit, maybe run `go fmt`.

File: internal/provider/resource_access_list_test.go [hunk 3]
```diff
@@ -40,7 +47,7 @@ func TestTimeUnmarshal(t *testing.T) {
 	//msg := `{"lastUpdateDateTime":"2021-08-03T15:20:29Z"}`
 	bodyBytes := []byte(msg)
 	type TestStruct struct{
-		LastUpdateDateTime *time.Time
+		LastUpdateDateTime *string
```
Comment (jeffbanks) [inline internal/provider/resource_access_list_test.go:50]: I recall this from a past PR for formatting.  I'm guessing the type changes are to ensure we don't bet that same parse error?

**Review by jeffbanks — COMMENTED** [review]
Just a few comments and questions.

File: internal/provider/resource_access_list.go [hunk 4]
```diff
@@ -73,39 +73,38 @@ func resourceAccessListCreate(ctx context.Context, d *schema.ResourceData, meta
 	client := meta.(*astra.ClientWithResponses)
 
 	databaseID := d.Get("database_id").(string)
-	addresses := d.Get("addresses")
+	addresses := d.Get("addresses").([]interface{})
```
[the remark this answers is not in the document]
Comment (emerkle826) [reply internal/provider/resource_access_list.go:76]: Not currently as the plugin defines it as required. So Terraform wouldn't allow the entry if there was no "addresses" section.

File: internal/provider/resource_access_list.go [hunk 5]
```diff
@@ -73,39 +73,38 @@ func resourceAccessListCreate(ctx context.Context, d *schema.ResourceData, meta
...
+			addressList,
+		)
 
-	resp, err := client.AddAddressesToAccessListForDatabaseWithResponse(ctx,
-		astra.DatabaseIdParam(databaseID),
-		addressList,
-	)
+		if err != nil {
+			return diag.FromErr(err)
+		} else if resp.StatusCode() >= 400 {
```
[the remark this answers is not in the document]
Comment (emerkle826) [reply internal/provider/resource_access_list.go:99]: I think it is required here. The call to `AddAddressesToAccessListForDatabaseWithResponse` will set an error if something goes wrong making the call. However, the call could be "successful", but still have a 400 level response status code. So the `else` here is to cover the case where the client didn't have an error making the request, but the response indicates it didn't actually succeed on the server side.

File: internal/provider/resource_access_list.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (jeffbanks) [reply internal/provider/resource_access_list.go:76]: 👍🏻

File: internal/provider/resource_access_list_test.go (hunk 1 above)
Comment (emerkle826) [reply internal/provider/resource_access_list_test.go:29]: I could, but to be honest, these tests just do an ADD, then a READ, then a DELETE and make sure none of those operations fail. Also, the READ doesn't verify that what you added got added, nor does the DELETE verify that everything got removed. Testing in general needs some more attention....

File: internal/provider/resource_access_list_test.go (hunk 3 above)
Comment (emerkle826) [reply internal/provider/resource_access_list_test.go:50]: Yep. I didn't realize there were JSON parsing tests for this, so I am fixing them here.

File: internal/provider/resource_access_list.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (jeffbanks) [reply internal/provider/resource_access_list.go:99]: I am 👍🏻 with the else if, but if the error exists it should never hit the else as it performs a return.

File: internal/provider/resource_access_list_test.go (hunk 1 above)
Comment (jeffbanks) [reply internal/provider/resource_access_list_test.go:29]: 💯

File: internal/provider/resource_access_list_test.go (hunk 3 above)
Comment (jeffbanks) [reply internal/provider/resource_access_list_test.go:50]: 👍🏻

**Review by jeffbanks — APPROVED** [review]

File: internal/provider/resource_access_list.go (hunk 2 above)
Comment (emerkle826) [reply internal/provider/resource_access_list.go:141]: I did run it, but it didn't change anything. Opening this in a different editor shows me that the spacing is all messed up. I'll update that.

Comment (emerkle826) [reply internal/provider/resource_access_list.go:141]: I take it back, the tabs/spacing is the same as the rest. It looks funny for the bit of code that is commented out, but it's the same for the code changes.

**Pull request merged**